connect_ferc1_to_eia.py

```diff
@@ -486,6 +468,55 @@ def calc_wins(df, ferc1_options, iqr_perc_diff):
     return winners
 
 
+def override_winners_with_training_df(winners, train_df):
+    """
+    Override winning matches with training data matches.
+
+    We want to ensure that all of the matches that we put in the training
+    data for the record linkage model actually end up in the resutls from the
+    record linkage model.
+
+    Args:
+        winners (pandas.DataFrame): winning matches generated via `calc_wins`.
+            Matches that had the highest rank in their record_id_ferc1, by a
+            wide enough margin.
+        train_df (pandas.DataFrame): training data/known matches between ferc
+            and the master unit list. Result of `prep_train_connections`.
+
+    Returns:
+        pandas.DataFrame: overridden winning matches. Matches that show up in
+        the training data `train_df` or if there was no corresponding training
+        data, matches that had the highest rank in their record_id_ferc1, by a
+        wide enough margin.
+    """
+    # we want to override the eia when the training id is
+    # different than the "winning" match from the recrod linkage
+    winners = (
+        pd.merge(
+            winners.reset_index(),
+            train_df.reset_index(),
+            on=['record_id_ferc1'],
+            how='outer',
+            suffixes=('_rl', '_trn'))
+        .assign(
+            record_id_eia=lambda x: np.where(
+                x.record_id_eia_trn.notnull(),
+                x.record_id_eia_trn,
+                x.record_id_eia_rl)
+        )
+    )
+    # check how many records were overridden
+    overridden = winners.loc[
+        (winners.record_id_eia_trn.notnull())
+        & (winners.record_id_eia_rl.notnull())
+        & (winners.record_id_eia_trn != winners.record_id_eia_rl)
+    ]
+    logger.info(f"Overridden records: {len(overridden)}")
```

Comment:
Here might we want to know what fraction of the records that had both winners and training data ended up being overridden in addition to the absolute number? If it's a big percentage what would that mean about our training/override set?

Reply:
Good idea. If it is a big % than our model didn't do a good job. And we can know that through some of the scores that we get from the model tuning. The thing I have been working on after this PR includes pulling this module together and I've including a warning if the best model isn't doing a great job.